pallets/block-rewards/src/migrations.rs

```diff
@@ -0,0 +1,159 @@
+#[cfg(feature = "try-runtime")]
+use {
+	cfg_traits::rewards::AccountRewards,
+	codec::{Decode, Encode},
+	num_traits::Zero,
+};
+
+use crate::{ActiveSessionData, Config, Pallet, SessionData};
+
+pub struct InitBlockRewards<T, CollatorReward, TotalReward>(
+	PhantomData<(T, CollatorReward, TotalReward)>,
+);
+
+fn get_collators<T: pallet_collator_selection::Config>() -> Vec<T::AccountId> {
+	let candidates = BoundedVec::<
+		T::AccountId,
+		<T as pallet_collator_selection::Config>::MaxCandidates,
+	>::truncate_from(
+		pallet_collator_selection::Pallet::<T>::candidates()
+			.into_iter()
+			.map(|c| c.who)
+			.collect(),
+	);
+	pallet_collator_selection::Pallet::<T>::assemble_collators(candidates)
+}
+
+impl<T, CollatorReward, TotalReward> OnRuntimeUpgrade
+	for InitBlockRewards<T, CollatorReward, TotalReward>
+where
+	T: frame_system::Config + Config + pallet_collator_selection::Config,
+	<T as Config>::Balance: From<u128>,
+	CollatorReward: Get<u128>,
+	TotalReward: Get<u128>,
+{
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+		assert_eq!(
+			Pallet::<T>::on_chain_storage_version(),
+			StorageVersion::new(0),
+			"On-chain storage version should be 0 (default)"
+		);
+		let collators = get_collators::<T>();
+		assert!(!collators.is_empty());
+
+		assert!(!CollatorReward::get().is_zero());
+		assert!(
+			TotalReward::get()
+				>= CollatorReward::get().saturating_mul(collators.len().saturated_into())
+		);
+
+		log::info!("💰 BlockRewards: Pre migration checks successful");
+
+		Ok(collators.encode())
+	}
+
+	// Weight: 2 + collator_count reads and writes
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		if Pallet::<T>::on_chain_storage_version() == StorageVersion::new(0) {
+			log::info!("💰 BlockRewards: Initiating migration");
+			let mut weight: Weight = Weight::zero();
+
+			let collators = get_collators::<T>();
+			weight.saturating_accrue(T::DbWeight::get().reads(2));
+
+			<T as Config>::Rewards::attach_currency(
+				(
+					<T as Config>::Domain::get(),
+					<T as Config>::StakeCurrency::get(),
+				),
+				COLLATOR_GROUP_ID,
+			)
+			.map_err(|e| log::error!("Failed to attach currency to collator group: {:?}", e))
+			.ok();
+
+			ActiveSessionData::<T>::set(SessionData::<T> {
+				collator_reward: CollatorReward::get().into(),
+				total_reward: TotalReward::get().into(),
+				collator_count: collators.len().saturated_into(),
+			});
+			weight.saturating_accrue(T::DbWeight::get().writes(1));
+
+			for collator in collators.iter() {
+				// TODO: Benching preferred to be precise.
+				// However, not necessarily needed as num of collators <= 10.
+				Pallet::<T>::do_init_collator(collator)
+					.map_err(|e| {
+						log::error!("Failed to init genesis collators for rewards: {:?}", e);
+					})
+					.ok();
+				weight.saturating_accrue(T::DbWeight::get().reads_writes(6, 6));
+			}
+			Pallet::<T>::current_storage_version().put::<Pallet<T>>();
```

Comment:
Is this substrate magic? Where is the version defined?

Reply:
The function itself [is part of the](https://paritytech.github.io/substrate/master/frame_support/dispatch/trait.GetStorageVersion.html#tymethod.current_storage_version) `GetStorageVersion` trait. The pallet's storage version `current_storage_version` [is in scope](https://github.com/centrifuge/centrifuge-chain/blob/feat/impl-block-rewards/pallets/block-rewards/src/lib.rs#L120) of the `frame_support::pallet` magic. 

So anytime we bump a pallet's storage version we still need to do a migration to have the `on_chain_storage_version` and `current_storage_version` aligned.